### DISCUSSION TITLE
flymake-show-project-diagnotics not updating (eglot for Java with jdtls)

### DISCUSSION THREAD

**Discussion by davidaventimiglia-professional** [joaotavora/eglot#1131] [Q&A]
Hello,

When I run `flymake-show-project-diagnostics` some warnings are listed for my Java project.  As I jump to those locations and fix the warnings, the diagnostics buffer doesn't update.  One problem I have is that I don't even know where to begin to look in order to troubleshoot this.  It could be a problem with eglot, jdtls, flymake, or some combination thereof.  Any suggestions?

**Comment by joaotavora** [comment]
Hmm, yes, it's hard to tell. Here's a suggestion: as you fix the warnings/errors check in the Eglot events buffer if the server is sending an updated list of diagnostics `publishDiagnostics` for the file you are editing. It should contain all the errors minus the one you've fixed. Report back here with the results of that experiment.

**Reply by davidaventimiglia-professional** [reply]
Thanks!  To keep things simple, one category of diagnostics that's giving me trouble is of the "The import ... is never used" variety in Java.  Taking an example Java servlet file `ProtonServlet.java`, when I add an unused import there's an event like this:

```lisp
[server-notification] Sun Dec  4 12:01:33 2022:
(:jsonrpc "2.0" :method "textDocument/publishDiagnostics" :params
	  (:uri "file:///home/neptunestationorg/Work/GitHub/ProtonChamber/src/main/java/org/protonchamber/ProtonServlet.java" :diagnostics
		[(:range
		  (:start
		   (:line 3 :character 7)
		   :end
		   (:line 3 :character 15))
		  :severity 2 :code "268435844" :source "Java" :message "The import java.sql is never used" :tags
		  [1])]))
```

That makes sense.  When I delete the unused import, there's an event like this:

```lisp
[server-notification] Sun Dec  4 12:02:03 2022:
(:jsonrpc "2.0" :method "textDocument/publishDiagnostics" :params
	  (:uri "file:///home/neptunestationorg/Work/GitHub/ProtonChamber/src/main/java/org/protonchamber/ProtonServlet.java" :diagnostics
		[]))
```

That also makes sense, as it seems to be that the server is publishing an empty list of diagnostics for this file.  Nevertheless, the `flymake-show-project-diagnostics` buffer still has this entry:

```
ProtonServlet.java            4   8 warning  n        Java [268435844]: The import java.sql is never used
```

Possibly, this rules out the jdtls server so that the problem is in eglot and/or flymake?

**Reply by joaotavora** [reply]
Yes, that works seem to be the case: one of those two, and I'm leaning flymake. Please submit this as a bug report to bug-gnu-emacs@gnu.org with all this information, a reference to this discussion, and remember to CC me in the report.

**Reply by davidaventimiglia-professional** [reply]
You bet.  Thanks!

**Reply by davidaventimiglia-professional** [reply]
Aha!  I filed a bug report, but if you'll forgive this discussion continuing a little longer, I see that _sometimes_ warnings are supplied by one back-end, but then when I fix the issue the same warning is then provided by a different back-end.  Here's what I mean.  I have another file with an unused import for `java.sql`.  When I do, the diagnostics buffer shows an entry from the `e-f-b` back-end, which I take to be the "eglot-flymake-backend":

```
SQLServlet.java               4   7 warning  e-f-b    Java [268435844]: The import java.sql is never used
```

If I delete the offending line, the diagnostics buffer replaces that with this:

```
SQLServlet.java               4   8 warning  n        Java [268435844]: The import java.sql is never used
```

I don't know what the `n` back-end is, but evidently it's publishing its own diagnostic for this code `268435844` which somehow is shadowed by the same diagnostic for the same code from the `e-f-b` back-end.  When the `e-f-b` back-end correctly publishes an empty diagnostic report for this file, the diagnostic report from the `n` back-end becomes revealed.  Or something like that.

**Comment by joaotavora** [comment]
David, please share this information in the ensuing bug report.  Let's
discuss this only in one place:
https://debbugs.gnu.org/cgi/bugreport.cgi?bug=59824

The `n` backend is probably the `nil` backend, which would be a bug in
itself.  I think I've fixed the problem in the meantime, so it may not
matter.

On Sun, Dec 4, 2022 at 8:53 PM David A. Ventimiglia <
***@***.***> wrote:

> Aha! I filed a bug report, but if you'll forgive this discussion
> continuing a little longer, I see that *sometimes* warnings are supplied
> by one back-end, but then when I fix the issue the same warning is then
> provided by a different back-end. Here's what I mean. I have another file
> with an unused import for java.sql. When I do, the diagnostics buffer
> shows an entry from the e-f-b back-end, which I take to be the
> "eglot-flymake-backend":
>
> SQLServlet.java               4   7 warning  e-f-b    Java [268435844]: The import java.sql is never used
>
> If I delete the offending line, the diagnostics buffer replaces that with
> this:
>
> SQLServlet.java               4   8 warning  n        Java [268435844]: The import java.sql is never used
>
> I don't know what the n back-end is, but evidently it's publishing its
> own diagnostic for this code 268435844 which somehow is shadowed by the
> same diagnostic for the same code from the e-f-b back-end. When the e-f-b
> back-end correctly publishes an empty diagnostic report for this file, the
> diagnostic report from the n back-end becomes revealed. Or something like
> that.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/joaotavora/eglot/discussions/1131#discussioncomment-4308068>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAC6PQ4BW3Z4FHGREE7HHETWLUAFVANCNFSM6AAAAAASTQTP7Q>
> .
> You are receiving this because you commented.Message ID:
> ***@***.***>
>


-- 
João Távora

**Reply by davidaventimiglia-professional** [reply]
done
